File: app/src/main/java/pl/tysia/maggstone/ui/presentation_logic/adapter/ErrorsAdapter.java
package pl.tysia.maggstone.ui.presentation_logic.adapter;


import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import pl.tysia.maggstone.R;
import pl.tysia.maggstone.data.model.Error;
import pl.tysia.maggstone.data.service.QueueItem;
import pl.tysia.maggstone.ui.DownloadStateActivity;

public class ErrorsAdapter extends CatalogAdapter<Error, ErrorsAdapter.SendingViewHolder> {
    ErrorsListener listener;

    public ErrorsAdapter(ArrayList<Error> items) {
        super(items);

    }

    public void setListener(ErrorsListener listener) {
        this.listener = listener;
    }

    public class SendingViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView title;
        TextView description;
        ImageView image;
        ImageButton forwardButton;
        ImageButton removeButton;

        private Observer<Integer> observer;

        SendingViewHolder(View v) {
            super(v);

            title = v.findViewById(R.id.title_tv);
            description = v.findViewById(R.id.description_tv);
            image = v.findViewById(R.id.icon_iv);
            forwardButton = v.findViewById(R.id.forward_button);
            removeButton = v.findViewById(R.id.remove_button);

            forwardButton.setOnClickListener(this);

            removeButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onErrorRemoveClicked(allItems.get(getAdapterPosition()));
                }
            });
        }

        @Override
        public void onClick(View v) {
            Context context = title.getContext();

            int pos = getAdapterPosition();

            switch (allItems.get(pos).getType()){
                case Error.TYPE_DOWNLOAD :
                    context.startActivity(new Intent(context, DownloadStateActivity.class));
                    break;
                case Error.TYPE_PICTURE :
                    if (listener != null) listener.onPictureSendClicked(allItems.get(pos));
                    break;
            }
        }

    }

    @NonNull
    @Override
    public SendingViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View v = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.list_item_error, viewGroup, false);

        v.setLayoutParams(new RecyclerView.LayoutParams(RecyclerView.LayoutParams.MATCH_PARENT, RecyclerView.LayoutParams.WRAP_CONTENT));

        SendingViewHolder vh = new SendingViewHolder(v);
        return vh;
    }

    @Override
    public void onBindViewHolder(@NonNull SendingViewHolder catalogItemViewHolder, int i) {
        Error item = shownItems.get(i);

        SendingViewHolder wareViewHolder = (SendingViewHolder) catalogItemViewHolder;

        wareViewHolder.title.setText(item.getTitle());
        wareViewHolder.description.setText(item.getDescription());

        switch (item.getType()){
            case Error.TYPE_CONNECTION :
                wareViewHolder.image.setImageResource(R.drawable.ic_round_wifi_off);
                wareViewHolder.forwardButton.setVisibility(View.GONE);
                wareViewHolder.removeButton.setVisibility(View.GONE);
                break;
            case Error.TYPE_DOWNLOAD :
                wareViewHolder.image.setImageResource(R.drawable.ic_round_sync_failed);
                wareViewHolder.forwardButton.setVisibility(View.VISIBLE);
                wareViewHolder.removeButton.setVisibility(View.GONE);
                break;
            case Error.TYPE_PICTURE :
                wareViewHolder.image.setImageResource(R.drawable.ic_round_image_not_sent);
                wareViewHolder.forwardButton.setVisibility(View.VISIBLE);
                wareViewHolder.removeButton.setVisibility(View.VISIBLE);
                break;
        }

    }

    @Override
    public void onViewDetachedFromWindow(@NonNull SendingViewHolder holder) {
        super.onViewDetachedFromWindow(holder);

    }

    public interface ErrorsListener{
        void onPictureSendClicked(Error error);

        void onErrorRemoveClicked(Error error);
    }
}

File: app/src/main/java/pl/tysia/maggstone/ui/presentation_logic/adapter/ErrorsAdapter.kt
package pl.tysia.maggstone.ui.presentation_logic.adapter

import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import pl.tysia.maggstone.R
import pl.tysia.maggstone.data.model.Error
import pl.tysia.maggstone.ui.DownloadStateActivity
import java.util.*

class ErrorsAdapter(items: ArrayList<Error>) :
    CatalogAdapter<Error, ErrorsAdapter.SendingViewHolder>(items) {

    var listener: ErrorsListener? = null

    inner class SendingViewHolder internal constructor(v: View) :
        RecyclerView.ViewHolder(v), View.OnClickListener {

        var title: TextView = v.findViewById(R.id.title_tv)
        var description: TextView = v.findViewById(R.id.subtitle_tv)
        var image: ImageView = v.findViewById(R.id.icon_iv)
        var forwardButton: ImageButton = v.findViewById(R.id.forward_button)
        var removeButton: ImageButton = v.findViewById(R.id.remove_button)

        private val observer: Observer<Int>? = null

        override fun onClick(v: View) {
            val context = title.context
            val pos = adapterPosition
            when (allItems[pos]!!.type) {
                Error.TYPE_DOWNLOAD -> context.startActivity(
                    Intent(
                        context,
                        DownloadStateActivity::class.java
                    )
                )
                Error.TYPE_PICTURE -> if (listener != null) listener!!.onPictureSendClicked(
                    allItems[pos]
                )
            }
        }

        init {
            forwardButton.setOnClickListener(this)
            removeButton.setOnClickListener {
                listener!!.onErrorRemoveClicked(
                    allItems[adapterPosition]
                )
            }
        }
    }

    override fun onCreateViewHolder(
        viewGroup: ViewGroup,
        i: Int
    ): SendingViewHolder {

        val v = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.list_item_error, viewGroup, false)

        v.layoutParams = RecyclerView.LayoutParams(
            RecyclerView.LayoutParams.MATCH_PARENT,
            RecyclerView.LayoutParams.WRAP_CONTENT
        )
        return SendingViewHolder(v)
    }

    override fun onBindViewHolder(
        catalogItemViewHolder: SendingViewHolder,
        i: Int
    ) {

        val item = shownItems[i]!!

        val wareViewHolder =
            catalogItemViewHolder

        wareViewHolder.title.text = item.getTitle()
        wareViewHolder.description.text = item.getDescription()

        when (item.type) {
            Error.TYPE_CONNECTION -> {
                wareViewHolder.image.setImageResource(R.drawable.ic_round_wifi_off)
                wareViewHolder.forwardButton.visibility = View.GONE
                wareViewHolder.removeButton.visibility = View.GONE
            }
            Error.TYPE_DOWNLOAD -> {
                wareViewHolder.image.setImageResource(R.drawable.ic_round_sync_failed)
                wareViewHolder.forwardButton.visibility = View.VISIBLE
                wareViewHolder.removeButton.visibility = View.GONE
            }
            Error.TYPE_PICTURE -> {
                wareViewHolder.image.setImageResource(R.drawable.ic_round_image_not_sent)
                wareViewHolder.forwardButton.visibility = View.VISIBLE
                wareViewHolder.removeButton.visibility = View.VISIBLE
            }
        }
    }

    override fun onViewDetachedFromWindow(holder: SendingViewHolder) {
        super.onViewDetachedFromWindow(holder)
    }

    interface ErrorsListener {
        fun onPictureSendClicked(error: Error)
        fun onErrorRemoveClicked(error: Error)
    }
}